Cubical/Foundations/EquivCharacterization.agda

```diff
@@ -0,0 +1,61 @@
+{-
+
+More theory about equivalences/different characterizations of equivalence
+
+- 'cong f' is an equivalence, if f is an equivalence
+
+-}
+{-# OPTIONS --cubical --safe #-}
+
+module Cubical.Foundations.EquivCharacterization where
+
+open import Cubical.Core.Everything
+
+open import Cubical.Foundations.Univalence
+open import Cubical.Foundations.Equiv
+
+isEquivCong : ∀ {ℓ} {A B : Set ℓ} {x y : A} (e : A ≃ B) → isEquiv (λ (p : x ≡ y) → (cong (e .fst) p))
+isEquivCong e = EquivJ (λ (B' A' : Set _) (e' : A' ≃ B') →
+                         (x' y' : A') → isEquiv (λ (p : x' ≡ y') → cong (e' .fst) p))
+                       (λ _ x' y' → idIsEquiv (x' ≡ y')) _ _ e _ _
+
+congEquiv : ∀ {ℓ} {A B : Set ℓ} {x y : A} (e : A ≃ B) → (x ≡ y) ≃ (e .fst x ≡ e .fst y)
+congEquiv e = ((λ (p : _ ≡ _) → cong (e .fst) p) , isEquivCong e)
+
+{-
+  Everything about path split equivalences is from https://arxiv.org/abs/1706.07526
+-}
+record isPathSplitEquiv {ℓ} {A B : Set  ℓ} (f : A → B) : Set ℓ where
+  field
+    s : B → A 
+    isSection : section f s
+    isSectionCong : (x y : A) → Σ[ s' ∈ (f(x) ≡ f(y) → x ≡ y) ] section (cong f) s'
+
+idIsPathSplitEquiv : ∀ {ℓ} {A : Set ℓ} → isPathSplitEquiv (λ (x : A) → x)
+idIsPathSplitEquiv = record {
+                       s = λ x → x ;
+                       isSection = λ x → refl ;
+                       isSectionCong = λ x y → (λ p → p) , λ p _ → p
+                     }
+
+module _ {ℓ} {A B : Set ℓ} where
+  toIsEquiv : (f : A → B) → isPathSplitEquiv f → isEquiv f 
+  toIsEquiv f record { s = r ; isSection = isSection ; isSectionCong = isSectionCong } =
+    (isoToEquiv f r isSection λ (x : A) → (isSectionCong (r (f x)) x).fst (isSection (f x))) .snd
+
+  sectionOfEquiv : {A' B' : Set ℓ} → (f : A' → B') → isEquiv f → B' → A'
+  sectionOfEquiv f record { equiv-proof = all-fibers-contractible } x =
+    all-fibers-contractible x .fst .fst
+
+  isSection : {A' B' : Set ℓ} → (f : A' → B') → (pf : isEquiv f) → section f (sectionOfEquiv f pf)
+  isSection f record { equiv-proof = all-fibers-contractible } =
+    λ x → all-fibers-contractible x .fst .snd 
+
+  fromIsEquiv : (f : A → B) → isEquiv f → isPathSplitEquiv f
+  fromIsEquiv f pf = 
+    record {
```

Comment:
And guess what... please use copatterns :)

Reply:
But this could be done with copatterns?

Reply:
Yes. I forgot to commit. I was trying to figure out, what makes it so slow to type check.

Reply:
Ah. No. It is in my last commit.

Reply:
What is slow to typecheck?

Reply:
The 'fromIsEquiv' takes something like 5 seconds. I made everything explicit now, but it is still slow.

Reply:
What's slow is checking that the inferred type of `isSec cong-f eq-cong` is the same as the expected type. And it's slow because Agda ends up unraveling `sectionOfEquiv (λ p i → f (p i)) eq-cong` to its WHNF on both sides when comparing the two types.

Because of `eq-cong` the terms being compared get quite large, hence the slowness.

Ironically, the two types being compared are pretty much the same, but Agda still does not recognize them as "syntactically equal", so they get reduced.

Reply:
How can one fix this Saizan? Put something in a record with a single constructor (this seems to always be the answer to this question...)?

Reply:
In this case we can just make sure eq-cong doesn't show up in types that get compared:

```agda
  sectionOfEquiv' : {A' B' : Set ℓ} → (f : A' → B') → isEquiv f → B' → A'
  sectionOfEquiv' f record { equiv-proof = all-fibers-contractible } x =
    all-fibers-contractible x .fst .fst

  isSec : {A' B' : Set ℓ} → (f : A' → B') → (pf : isEquiv f) → section f (sectionOfEquiv' f pf)
  isSec f record { equiv-proof = all-fibers-contractible } x =
    all-fibers-contractible x .fst .snd

  sectionOfEquiv : {A' B' : Set ℓ} → (f : A' → B') → isEquiv f → Σ (B' → A') (section f)
  sectionOfEquiv f e = sectionOfEquiv' f e , isSec f e

  fromIsEquiv : (f : A → B) → isEquiv f → isPathSplitEquiv f
  s (fromIsEquiv f pf) = sectionOfEquiv' f pf
  sec (fromIsEquiv f pf) = isSec f pf
  secCong (fromIsEquiv f pf) x y = sectionOfEquiv cong-f eq-cong
	where
          cong-f : x ≡ y → f x ≡ f y
          cong-f = λ (p : x ≡ y) → cong f p
          eq-cong : isEquiv cong-f
          eq-cong = isEquivCong (f , pf)
```

Reply:
Thanks for fixing it and explaining:)